### DISCUSSION TITLE
Will jQuery.event.special[EVENT_NAME].trigger be removed from jQuery 4?

### DISCUSSION THREAD

**Discussion by effulgentsia** [jquery/jquery#5172] [Q&A] [answered]
https://github.com/jquery/jquery/wiki/jQuery-4.0-Event-Design#remove-special-events-hooks says that special event hooks will be removed from jQuery 4. However, `jQuery.event.special[EVENT_NAME].trigger` [has not yet been removed from the main branch](https://github.com/jquery/jquery/blob/main/src/event/trigger.js#L68), searching for "event.special" doesn't turn up any issues or PRs since 2015, and the [jQuery 4 milestone](https://github.com/jquery/jquery/milestone/7) seems pretty close.

Is it reasonable to expect `jQuery.event.special[EVENT_NAME].trigger` to continue to work in jQuery 4? I'm asking because we're [considering using it in an upcoming version of Drupal](https://www.drupal.org/project/drupal/issues/3324062#comment-14806384), and are wondering if that'll create problems for us when upgrading to jQuery 4.

Thanks for any insight you can offer.

**Comment by mgol** [comment, marked as answer]
We’ve changed our plans and we’re now targeting the event module rewrite at jQuery 5.0, not 4.0. We don’t plan to remove this API in 4.0. I’ll update that wiki page, thanks for the reminder!

We’re now focused on jQuery 3.6.2 & 3.7.0 but we should start working on an upgrade guide for jQuery 4.0.0 soon. That should help you with planning the upgrade.

EDIT: I edited the page, it's now at [jQuery 5.0 Event Design](https://github.com/jquery/jquery/wiki/jQuery-5.0-Event-Design).

**Comment by jun1432** [comment]
Hi mgol, I just updated recently on 3.7.1 but stuck on the error of jQuery.event.special on chrome with toggle device toolbar for responsive view. Is there any chance of a fix for this? Exact error below:

`Uncaught TypeError: ((jQuery.event.special[handleObj.origType] || {}).handle || handleObj.handler).apply is not a function`

I tried defining the touch events:
$.each(["touchstart", "touchmove", "touchend"], function (i, eventName) {
    $.event.special[eventName] = $.event.special[eventName] || {};
});

The issue still persists.

**Reply by mgol** [reply]
If you found an issue, please construct a runnable test case and report an issue. We cannot debug it for yourself just from an error snippet, that's too little data.
